fuzz/ml-kem.c

```diff
@@ -0,0 +1,685 @@
+ * @param len       Pointer to the remaining buffer size; updated accordingly.
+ * @param keytype   Pointer to store the selected key type string.
+ * @param keylen    Pointer to store the selected key length.
+ * @param only_valid Flag to restrict selection to valid key sizes.
+ *
+ * @return 1 if a key type is successfully selected, 0 on failure.
+ */
+static int select_keytype_and_size(uint8_t **buf, size_t *len,
+                                   char **keytype, size_t *keylen,
+                                   int only_valid)
+{
+    uint16_t keysize;
+    uint16_t modulus = 6;
+
+    *buf = consume_uint16t(*buf, len, &keysize);
+    if (*buf == NULL)
+        return 0;
+
+    /*
+     * select from sizes
+     * ML-KEM-512, ML-KEM-768, and ML-KEM-1024
+     * also select some invalid sizes to trigger
+     * error paths
+     */
+    if (only_valid)
+        modulus = 3;
+
+    /*
+     * Note, keylens for valid values (cases 0-2)
+     * are taken based on input values from our unit tests
+     */
+    switch (keysize % modulus) {
+    case 0:
+        *keytype = "ML-KEM-512";
+        *keylen = 800;
+        break;
+    case 1:
+        *keytype = "ML-KEM-768";
+        *keylen = 1172;
+        break;
+    case 2:
+        *keytype = "ML-KEM-1024";
+        *keylen = 1568;
+        break;
+    case 3:
+        /* select invalid alg */
+        *keytype = "ML-KEM-13";
+        *keylen = 13;
+        break;
+    case 4:
+        /* Select valid alg, but bogus size */
+        *keytype = "ML-KEM-1024";
+        *buf = consume_uint16t(*buf, len, &keysize);
+        *keylen = (size_t)keysize;
+        *keylen %= 2048; /* size to our key buffer */
+        break;
+    default:
+        *keytype = NULL;
+        *keylen = 0;
+        break;
+    }
+    return 1;
+}
+
+/**
+ * @brief Creates an ML-KEM raw key from a buffer.
+ *
+ * This function selects a key type and size from the buffer, generates
+ * a random key of the appropriate length, and creates either a public
+ * or private ML-KEM key using OpenSSL's EVP_PKEY interface.
+ *
+ * @param buf   Pointer to the buffer pointer; updated after reading.
+ * @param len   Pointer to the remaining buffer size; updated accordingly.
+ * @param key1  Pointer to store the generated EVP_PKEY key (public or private).
+ * @param key2  Unused parameter (reserved for future use).
+ *
+ * @note The generated key is allocated using OpenSSL's EVP_PKEY functions
+ *       and should be freed appropriately using `EVP_PKEY_free()`.
+ */
+static void create_mlkem_raw_key(uint8_t **buf, size_t *len,
+                                 void **key1, void **key2)
+{
+    EVP_PKEY *pubkey;
+    char *keytype = NULL;
+    size_t keylen = 0;
+    uint8_t key[2048];
+    int pub = 0;
+
+    if (!select_keytype_and_size(buf, len, &keytype, &keylen, 0))
+        return;
+
+    /*
+     * Select public or private key creation based on the low order
+     * bit of the next buffer value
+     */
+    if ((*buf)[0] & 0x1)
```

Comment:
This won't work very well, unless when private is chosen, you double the length and add 32, but then a 2k sized buffer is not enough, since the private key for `ML-KEM-1024` is $`2*1568 + 32 =3,168`$ bytes long.